ftw/contentpage/browser/newslisting.py

```diff
@@ -67,20 +69,41 @@ def get_portlet(self):
         if not manager_name or not name:
             return
 
-        manager = getUtility(
-            IPortletManager,
-            name=manager_name,
-            context=self.context)
-        assignments = getMultiAdapter(
-            (self.context, manager),
-            IPortletAssignmentMapping,
-            context=self.context)
-
-        if name in assignments:
-            return queryMultiAdapter(
-                (self.context, self.request, self,
-                 manager, assignments[name]),
-                IPortletRenderer)
+        context = self.context
+
+        # Prepare a list of objects by walking up the path.
+        contexts = [context]
+        while not IPloneSiteRoot.providedBy(context):
+            if IAcquirer.providedBy(context):
+                context = aq_parent(aq_inner(context))
+            else:
+                context = context.__parent__
+            contexts.append(context)
+
+        # Prepare a list of tuples in the form `(manager, assignments)`.
+        managers_and_assignments = []
+        for context in contexts:
+            manager = getUtility(
+                IPortletManager,
+                name=manager_name,
+                context=context)
+            assignments = getMultiAdapter(
+                (context, manager),
+                IPortletAssignmentMapping,
+                context=context)
+
+            if assignments is not None:
+                managers_and_assignments.append((manager, assignments))
```

Comment:
I'd probably move building the managers / assignments list into a separate helper method in order to reduce complexity of from this method :wink: